torch/csrc/distributed/rpc/py_rref.cpp

```diff
@@ -0,0 +1,102 @@
+#include <torch/csrc/distributed/rpc/py_rref.h>
+
+#include <torch/csrc/distributed/rpc/rref_context.h>
+#include <torch/csrc/jit/pybind_utils.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+namespace {
+
+// Constants below are used in PyRRef pickling and unpickling. PyRRef is
+// converted into a py::tuple in pickling, and reconstructed from the py::tuple
+// in pickling.
+constexpr int RREF_TUPLE_SIZE = 2; // number of data fields in the py::tuple
+constexpr int RFD_IDX = 0; // index of RRefForkData
+constexpr int TYPE_IDX = 1; // index of type (py::object or IValue)
+
+} // namespace
+
+PyRRef::PyRRef(std::shared_ptr<RRef> rref) : rref_(std::move(rref)) {
+  TORCH_CHECK(rref_, "PyRRef must not wrap nullptr");
+}
+
+bool PyRRef::isOwner() const {
+  return rref_->isOwner();
+}
+
+worker_id_t PyRRef::owner() const {
+  return rref_->owner();
+}
+
+py::object PyRRef::toHere() {
+  if (rref_->isPyObj()) {
+    if (rref_->isOwner()) {
+      auto& value =
+          std::static_pointer_cast<OwnerRRef<py::object>>(rref_)->getValue();
+      AutoGIL ag;
```

Comment:
> Is there an implicit conversion happening upon returning?

Yes, the return statement converts a `py::object` reference to a `py::object`. `getValue()` returns a const reference and `setValue(T&&)` takes a rvalue argument to avoid acquiring GIL in them. Any caller that needs to get a copy or make a copy should acquire GIL themselves. Let me make this explicit.